### PR TITLE
uctypes: Implement assignment for scalar array

### LINKED ISSUE #1743: uctypes: assign to scalar in an array is not fully implemented
```python
import uctypes

desc = {
    # arr is array at offset 0, of UINT32 elements, array size is 2
    "arr": (uctypes.ARRAY | 0, uctypes.UINT32 | 2),
}

data = bytearray(8)
s = uctypes.struct(uctypes.addressof(data), desc)

s.arr[0] = 0x11223344
```

gives ```TypeError: 'struct' object does not support item assignment```

It only works for UINT8 (because the array comes out as a bytearray).

I've got a patch+test for this, but #1698 mention I big refactor coming.
Are you interested in it anyway ?

### PR DESCRIPTION
**Pull request by Anton-2**
Fix #1743

There is still some weirdness when doing unaligned access with a native layout. Data seems to be written with the wrong endianness.

The corresponding test case at the end of tests/extmod/uctypes_array_assign_native.py are commented out. I dont know how to fix it, and it may be related to #1698, needing more deeper changes.

### CODE REVIEW

**Comment by Anton-2** [comment]
Slight rework of #1745, casting pointer to uintptr_t instead of uint in IS_ALIGNED.

File: extmod/moductypes.c [hunk 1]
```diff
@@ -107,6 +107,9 @@ enum {
 // We cannot apply the below to INT8, as their range [-128, 127]
 #define IS_SCALAR_ARRAY_OF_BYTES(tuple_desc) (GET_TYPE(MP_OBJ_SMALL_INT_VALUE((tuple_desc)->items[1]), VAL_TYPE_BITS) == UINT8)
 
+#define IS_ALIGNED(addr, val_type) (!(((uintptr_t)addr) & (GET_SCALAR_SIZE(val_type)-1)))
```
Comment (pfalcon) [inline extmod/moductypes.c]: I don't think this having this macro is needed/a correct way. The rules are simple:if structure is native, you must use integral instructions to access values of given size (i.e., access 8, 16, 32, 64 bit quantities as such). Whether it's aligned or not is not important on this level - it's up to the designer of the structure, who should keep in mind that on many architecture, unaligned access to integral values lead to exceptions (so, if someone uses unaligned native structure on such CPU, exceptions is what they want, and thus, must get).

When using non-native structures, there's no constraint of integral access, any quantity may be accessed byte by byte, and alignment exceptions must not happen.

**Comment by pfalcon** [comment]
> tests/extmod/uctypes_array_assign_native.py

Testing of native structures is inherently hard - at the very least, you assume a particular byte order. That's why current native tests have "native_le.py" suffix and explicit check that they're indeed run on little-ending arch. (And if course they must use only aligned structures.)

**Comment by pfalcon** [comment]
So, besides those issues, it seems to be ok, and I'm going to fix them myself now, as there're more uctypes issues in queue.

**Comment by pfalcon** [comment]
Merged with the fixes above, thanks!

**Pull request closed without merging**